[doc] Add DataPipeline + Callbacks + Registry (#207)
PyTorchLightning/lightning-flash: docs/source/general/data.rst

Data
####

.. _data:

*******************************
Using DataModule + DataPipeline
*******************************

> 

Reviewer: This is missing context.

What are these objects? Why do they exist? when do I need to use them?

I think we can do something like
1. How to use out-of-the-box flashdatamodules
2. How to customize existing datamodules
3. How to build a datamodule for a new task
4. What are data pipelines and when are they required
5. how to use them

Author: Yes. Definitely. I will work on this next week :)

In this section, we will create a very simple ``ImageClassificationPreprocess`` with a ``ImageClassificationDataModule``.

Example::

    import os
    import numpy as np
    from flash.data.data_module import DataModule
    from flash.data.process import Preprocess
    from PIL.Image import Image
    import torchvision.transforms as T
    from torch import Tensor

    # Subclass ``Preprocess``

    class ImageClassificationPreprocess(Preprocess):

        to_tensor = T.ToTensor()

>         def load_data(self, folder: str, dataset: AutoDataset) -> Iterable:

Author: This is auto-injected and optional. If users works with Preprocess, the AutoDataset is handled for them in behind the scenes.
However, it can useful to get access to it to store metadata.

            # The AutoDataset is optional but can be useful to save some metadata.

>             # metadata looks like this: [(image_path_1, label_1), ... (image_path_n, label_n)].

Author: Done !

>             metadata = make_dataset_from_folder(folder)

Author: It was a dummy utils to explain how it works. But if you got confused, more will be. Replaced it by real function.

            # for the train ``AutoDataset``, we want to store the ``num_classes``.
            if self.training:
                dataset.num_classes = len(np.unique([m[1] for m in metadata]))

            return metadata

        def predict_load_data(self, predict_folder: str) -> Iterable:
            # This returns [image_path_1, ... image_path_m].
            return os.listdir(folder)

        def load_sample(self, sample: Union[str, Tuple[str, int]]) -> Tuple[Image, int]
            if self.predicting:

>                 return load_pil(image_path)

Author: Done !

            if self.predicting:
                return self.to_tensor(sample)
            else:
                return self.to_tensor(sample[0]), sample[1]

    class ImageClassificationDataModule(DataModule):

        # Set ``preprocess_cls`` with your custom ``preprocess``.

        preprocess_cls = ImageClassificationPreprocess

        @classmethod
        def from_folders(
            cls,
            train_folder: Optional[str],
            val_folder: Optional[str],
            test_folder: Optional[str],
            predict_folder: Optional[str],
            **kwargs
        ):

            preprocess = cls.preprocess_cls()

            # {stage}_load_data_input will be given to your
            # ``Preprocess`` ``{stage}_load_data`` function.
            return cls.from_load_data_inputs(
                    train_load_data_input=train_folder,
                    val_load_data_input=val_folder,
                    test_load_data_input=test_folder,
                    predict_load_data_input=predict_folder,
                    preprocess=preprocess,  # DON'T FORGET TO PASS THE CREATED PREPROCESS
                    **kwargs,
                )

>     dm = ImageClassificationDataModule.from_folders("./data/train", "./data/val", "./data/test", "./data/predict")

Author: Done !

Data
####

.. _data:

***********
Terminology
***********

Here are common terms you need to be familiar with:

.. list-table:: Terminology
   :widths: 20 80
   :header-rows: 1

   * - Term
     - Definition
   * - DataModule
     - The :class:`~flash.data.data_module.DataModule` contains the dataset, transforms and dataloaders.
   * - :class:`~flash.data.data_pipeline.DataPipeline`
     - The :class:`~flash.data.data_pipeline.DataPipeline` is Flash internal object to manage :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` objects.
   * - :class:`~flash.data.process.Preprocess`
     - The :class:`~flash.data.process.Preprocess` provides a simple hook-based API to encapsulate your pre-processing logic.

>         This enables to not rely on Dataset anymore even if supported.

Author: Updated the doc.

        Flash DataPipeline contains a system to call the right hooks when needed.
        The :class:`~flash.data.process.Preprocess` hooks covers from data-loading to model forwarding.
   * - :class:`~flash.data.process.Postprocess`
     - The :class:`~flash.data.process.Postprocess` provides a simple hook-based API to encapsulate your post-processing logic.
        The :class:`~flash.data.process.Postprocess` hooks covers from model outputs to predictions export.

*******************************************
How to use out-of-the-box flashdatamodules
*******************************************

Flash provides several DataModules with helpers functions.

> Checkout the ``Task Section`` to learn more about them.

Author: Yes, you commented out-dated doc.

> Checkout the ``Task Section`` to learn more about them.

Author: Updated this part.

********************************
Why Preprocess and PostProcess ?
********************************

> Currently, it is common practice to implement a :class:`~torch.data.utils.Dataset` and provide them to a :class:`~torch.data.utils.DataLoader`.

Author: Done !

> However, once the model is trained, lot of engineering work is required to enable the model

Author: Done !

to perform predictions on ``un-processed data`` (called raw data) or to study data transforms impact during training.

> But more importantly, it is hard to make a trainer model ready for production.

Author: Making inference on raw_data and putting the model in production are really different things, but I will remove it there.

> The :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` have been created to resolve those issues.

Author: Done !

> By providing a series of hooks which can overridden with custom data processing logic, the user have a much more granular control on the data processing flow.

Author: Done !

But it also makes your code more readable, modular and easy to extend.

To change the processing behavior only on specific stages for a given hook,

> you can prefix all the above hooks adding ``train``, ``val``, ``test`` or ``predict``.

Author: Done !

.. note::

    [WIP] We are currently working on a new feature to convert fully implemented :class:`~flash.data.process.Preprocess`

    and :class:`~flash.data.process.Postprocess` are fully implemented, to be fully deployable for ``Endpoints``

>     or ``BatchTransformJob`` directly from checkpoints. Stay tuned !

Author: Pushed update.

>     or ``BatchTransformJob`` directly from checkpoints. Stay tuned !

Reviewer 2: This is unclear. Will think of how to make it more clear.

*************************************
How to customize existing datamodules
*************************************

Currently, Flash Tasks are implementing using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess`.

> However, it is not a hard requirement and one can still use :class:`~torch.data.utils.Dataset`, but we highly recommend

Reviewer 2: One can still use :class:`~torch.data.utils.Dataset`, but ...

Author: ```
.. note::

    ``[WIP]`` We are currently working on a new feature to make :class:`~flash.data.process.Preprocess`

    and :class:`~flash.data.process.Postprocess` automatically deployable from checkpoints directly as

    ``Endpoints`` or ``BatchTransformJob``. Stay tuned !
```

using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` instead.

Example::

    from flash.data.data_module import DataModule

    dm = DataModule(train_dataset=MyDataset(train=True))
    trainer = Trainer(fast_dev_run=True)
    trainer.fit(model, data_module=dm)

In order to customize Flash to your need, you need to know what are :class:`~flash.data.data_module.DataModule`
and :class:`~flash.data.process.Preprocess` responsibilities.

.. note::

>     From this point, we strongly encourage the readers to quickly check the :class:`~flash.data.process.Preprocess` API before getting further.

Reviewer 2: At this point.

The :class:`~flash.data.data_module.DataModule` provides ``classmethod`` helpers to build
:class:`~flash.data.process.Preprocess` and :class:`~flash.data.data_pipeline.DataPipeline`,

> generate ``AutoDataset`` and populate DataLoaders from them.

Reviewer 2: What does it mean to generate AutoDataset?

Author: AutoDataset are internals class. Tried to clarify this.

The :class:`~flash.data.process.Preprocess` contains the processing logic related to a given task. One can easily override hooks

> to customize a built-in :class:`~flash.data.process.Preprocess` for his need.

Reviewer 2: their need

Author: Done !

Example::

    from flash.vision import ImageClassificationData, ImageClassifier, ImageClassificationPreprocess

    class CustomImageClassificationPreprocess(ImageClassificationPreprocess):

        @staticmethod
        def load_sample(sample) -> Tuple[Image.Image, int]:
            # By default, ``ImageClassificationPreprocess`` expects ``.PNG`` or ``.JPB`` to be loaded into PIL Image.
            # Assuming you have numpy image, just override this hook and add your own logic.
            numpy_image_path, label = sample
            return np.load(numpy_image_path), sample

    datamodule = ImageClassificationData.from_folders(
        train_folder="data/hymenoptera_data/train/",
        val_folder="data/hymenoptera_data/val/",
        test_folder="data/hymenoptera_data/test/",
        preprocess_cls=CustomImageClassificationPreprocess
    )

*****************************************************
How to build a Datamodule + Preprocess for a new task
*****************************************************

> In this section, we will create a very simple ``ImageClassificationPreprocess`` with a ``ImageClassificationDataModule``.

Reviewer 2: The example below shows a very simple ``ImageClassificationPreprocess`` with a ``ImageClassificationDataModule``.

Author: Done !

******************************

Preprocess
__________

.. note:: The ``load_data`` and ``load_sample`` will be used to generate an AutoDataset object.

Here is the ``AutoDataset`` pseudo-code.

Example::

    from pytorch_lightning.trainer.states import RunningStage

    class AutoDataset
        def __init__(
            self,
            data: Any,
            load_data: Optional[Callable] = None,
            load_sample: Optional[Callable] = None,
            data_pipeline: Optional['DataPipeline'] = None,
            running_stage: Optional[RunningStage] = None
        ) -> None:

            self.preprocess = data_pipeline._preprocess_pipeline
            self.preprocessed_data: Iterable = self.preprocess.load_data(data)

        def __getitem__(self, index):
            return self.preprocess.load_sample(self.preprocessed_data[index])

        def __len__(self):
            return len(self.preprocessed_data)

.. note::

    The ``pre_tensor_transform``, ``to_tensor_transform``, ``post_tensor_transform``, ``collate``,
    ``per_batch_transform`` are injected as the ``collate_fn`` function of the DataLoader.

Here is the pseudo code using the preprocess hooks name.

> Surely, Flash will take care of calling the right hooks for each stage.

Reviewer 2: Flash takes care of calling the right hooks for each stage.

Author: Done !

   * - :class:`~flash.data.data_pipeline.DataPipeline`
     - The :class:`~flash.data.data_pipeline.DataPipeline` is Flash internal object to manage :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` objects.
   * - :class:`~flash.data.process.Preprocess`
     - The :class:`~flash.data.process.Preprocess` provides a simple hook-based API to encapsulate your pre-processing logic.
        The :class:`~flash.data.process.Preprocess` provides multiple hooks such as :meth:`~flash.data.process.Preprocess.load_data`
        and :meth:`~flash.data.process.Preprocess.load_sample` which are used to replace a traditional `Dataset` logic.
        Flash DataPipeline contains a system to call the right hooks when needed.
        The :class:`~flash.data.process.Preprocess` hooks covers from data-loading to model forwarding.
   * - :class:`~flash.data.process.Postprocess`
     - The :class:`~flash.data.process.Postprocess` provides a simple hook-based API to encapsulate your post-processing logic.
        The :class:`~flash.data.process.Postprocess` hooks covers from model outputs to predictions export.

*******************************************

> How to use out-of-the-box flashdatamodules

Reviewer: Move this to the top of the page.

It should be 

**Using built-in DataModules**
flash has built in datamodules you can use to load your data to any existing flash task. For more info, check out....

**Custom Data**
If you are creating a new task or using a different type of dataset for an existing task, you may want to override .... 

and then add the terminology table and everything after that

*******************************************

Flash provides several DataModules with helpers functions.
Checkout the :ref:`image_classification` section or any other tasks to learn more about them.

********************************

> Why Preprocess and PostProcess ?

Reviewer: ```suggestion
Data Processing
```

Author: Done 1

********************************

Currently, it is common practice to implement a `Dataset <https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset>`_
and provide them to a `DataLoader <https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader>`_.

> However, after model training, it requires a lot of engineering overhead to make inference on raw data and deploy the model in production environnement.

Reviewer: Explain why? Since the pre+post processing steps are not included, etc

Author: Done !

> The :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` have been created to resolve these issues.

Reviewer: ```suggestion
The :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` classes can be used to store the data as well as the preprocessing and postprocessing transforms.
```

Author: Done !

The :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` have been created to resolve those issues.

> By providing a series of hooks which can overridden with custom data processing logic, the user have a much more granular control on his data processing flow.

Reviewer 3: ```suggestion
By providing a series of hooks that can be overridden with custom data processing logic, the user has much more granular control over their data processing flow.
```

Author: Done !

> But it also makes your code more readable, modular and easy to extend.

Reviewer: Maybe these can be bullet points:

Encapsulate the data processing transforms and the datasets is useful for:

* Making inference on raw data simple
* Make the code more readable, modular and self-contained
* 
anything else?

Author: Done !

To change the processing behavior only on specific stages for a given hook,
you can prefix each of the :class:`~flash.data.process.Preprocess` and  :class:`~flash.data.process.Postprocess`
hooks by adding ``train``, ``val``, ``test`` or ``predict``.

Check out :class:`~flash.data.process.Preprocess` for some examples.

.. note::

    ``[WIP]`` We are currently working on a new feature to make :class:`~flash.data.process.Preprocess`

    and :class:`~flash.data.process.Postprocess` automatically deployable from checkpoints as

    ``Endpoints`` or ``BatchTransformJob``. Stay tuned !

*************************************
How to customize existing datamodules
*************************************

> Currently, Flash Tasks are implementing using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess`.

Reviewer: ```suggestion
Currently, Flash Tasks are implemented using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess`.
```

Author: Done !

> Currently, Flash Tasks are implementing using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess`.

Reviewer: This can be a note AFTER explaining how to customize datamodules

to customize a built-in :class:`~flash.data.process.Preprocess` for their needs.

Example::

    from flash.vision import ImageClassificationData, ImageClassifier, ImageClassificationPreprocess

    class CustomImageClassificationPreprocess(ImageClassificationPreprocess):

        # Assuming you have images in numpy format,
        # just override ``load_sample`` hook and add your own logic.
        @staticmethod
        def load_sample(sample) -> Tuple[Image.Image, int]:
            # By default, ``ImageClassificationPreprocess`` expects
            # ``.png`` or ``.jpg`` to be loaded into PIL Image.
            numpy_image_path, label = sample
            return np.load(numpy_image_path), sample

    datamodule = ImageClassificationData.from_folders(
        train_folder="data/hymenoptera_data/train/",
        val_folder="data/hymenoptera_data/val/",
        test_folder="data/hymenoptera_data/test/",
        preprocess_cls=CustomImageClassificationPreprocess
    )

******************************
Custom Preprocess + Datamodule
******************************

> The example below shows a very simple ``ImageClassificationPreprocess`` with a ``ImageClassificationDataModule``.

Reviewer: Always start with the motivation. Start by explaining it in general, and then move to the example.

If you want to add custom bla and custom bla, you might need to create custom dm and a custom datapipeline.

1. Create a DataModule - The DataModule should have helper functions to instantiate the DataModule for a relevant input type (folder structure, csv, numpy array, etc). Let's say in our example we have folders of images arranged like.... then we will create a from_folders.... etc
2. Create the preprocess- your task might need some preprocess and post process transforms. In this example.... bla bla bla
3. Make sure the processing API is used in your datamodule
4. Use your new datamodule in your task
